rfc/005-soft-chain-upgrades.md

```diff
@@ -0,0 +1,473 @@
+internalized to allow for changes (so long as this wouldn't change the overall
+behavior).
+
+Another similar approach is to have the prior versions in a different repo and
+import it as a library. This may minimize the amount of code in the repo.
+
+### Current Use of Versioning
+
+Records of versions are currently stored in `State` and in the `Header` of each
+block. This already covers a broad base of scenarios. The consensus reactor can
+use the version from state to ensure it reads `Vote`s and `Partsetheader`s
+correctly. Similarly, fast sync, the block executor and evidence module, whom
+all interact with block protocol data structures also all have access to state.
+The next few sections will explore the aspects of Tendermint that will need to
+be modified to prepare for multi versioning support.
+
+### RPC
+
+Supporting soft upgrades means that the RPC will need to be able to deliver
+information across different versions. The RPC will therefore need to know which
+heights corresponds to which version and then be able to communicate this to
+the clients. This would most likely require clients to be versatile to handling
+the different block versions unless it was possible for the data to be
+transformed to an earlier version. For example, block explorers would also take
+on the burden of supporting all block versions across a chain. Clients would
+indicate the latest version they support in the RPC call itself e.g.
+`/v3/block?height=100`. The node would know what version the block at height 100
+was. If it was greater than v3 it would return a version error indicating the
+version the block explorer would need to support. If it was less or equal to v3
+then it would return the block (which contains version information).
+
+### Light Clients
+
+Similar to RPC, light clients will also need to be wary of block protocol
+changes and be able to handle them accordingly. Depending on these changes,
+there might be the need for special verification logic across signed headers
+with different versions. If a light client bisected to a height where it didn't
+have the version it would return an error.
+
+### Implications
+
+**Positive:**
+
+- It requires no changes to the database therefore keeps the immutable aspect
+  of the blockchain.
+- Async upgrading. Nodes choose when they want to upgrade the binary and
+Tendermint handles the actual transition.
+
+**Negative:**
+
+- Increased burden on RPC clients to be able to parse multiple versions.
+- Bloating of code base and binary. We would also need to figure out methods
+of maintaining good code hygiene with respect to having all these versions.
+- If we separate versions i.e. ABCI and Block and P2P this could potentially
+compound the problem if we need to consider the different permutations and
+how they can be supported. Correctness analysis and proper testing would be
+required to mitigate errors.
+
+
+## Method 2: Chain Migration
```

Comment:
I think this sounds pretty complex and I'm not sure it's worth it in the end, especially since it seems you either need a bunch more storage if you want to regenerate the original blocks or you basically need to rerun consensus for the whole blockchain or something? Also need to be careful if you're resigning things to make sure there's no risk of double signing.

Reply:
I feel like the big downside to the Multi Data Structure approach is that it puts some maintenance burden on clients and wallets as well. This is OK if they only need to support a handful of versions, but it could be prohibitively complex if we introduce a dozen versions or something